libnopegl/src/gen_specs.c

```diff
@@ -19,29 +19,156 @@
  * under the License.
  */
 
+#include <string.h>
+
 #include "hmap.h"
 #include "memory.h"
 #include "nopegl.h"
 #include "internal.h"
 #include "nodes_register.h"
+#include "params.h"
 
 #define CLASS_LIST(type_name, cls) extern const struct node_class cls;
 NODE_MAP_TYPE2CLASS(CLASS_LIST)
 
 extern const struct node_param ngli_base_node_params[];
 extern const struct param_specs ngli_params_specs[];
 
+#define REGISTER_NODE(type_name, cls)           \
+    case type_name: {                           \
+        extern const struct node_class cls;     \
+        ngli_assert(cls.id == type_name);       \
+        return &cls;                            \
+    }                                           \
+
+static const struct node_class *get_node_class(int type)
+{
+    switch (type) {
+        NODE_MAP_TYPE2CLASS(REGISTER_NODE)
+    }
+    return NULL;
+}
+
+static char *escape_json_str(const char *s)
+{
+    if (!s)
+        return NULL;
+    char *p = malloc(strlen(s) * 2 + 1);
+    if (!p)
+        return NULL;
+    char *ret = p;
+    while (*s) {
+        if (*s == '"')
+            *p++ = '\\';
+        *p++ = *s++;
+    }
+    *p = 0;
+    return ret;
+}
+
 static void print_node_params(const char *name, const struct node_param *p)
 {
     printf("  \"%s\": [\n", name);
     if (p) {
         while (p->key) {
-            printf("    [\"%s\", \"%s\", \"%s%s%s\"]%s\n", p->key, ngli_params_specs[p->type].name,
-                   (p->flags & NGLI_PARAM_FLAG_ALLOW_LIVE_CHANGE) ? "L" : "",  /* [L]ive */
-                   (p->flags & NGLI_PARAM_FLAG_ALLOW_NODE)        ? "N" : "",  /* [N]ode */
-                   (p->flags & NGLI_PARAM_FLAG_NON_NULL)          ? "M" : "",  /* [M]andatory */
-                   (p + 1)->key ? "," : ""
-            );
+            printf("    {\n");
+            printf("      \"name\": \"%s\",\n", p->key);
+            printf("      \"type\": \"%s\",\n", ngli_params_specs[p->type].name);
+#define D "      \"default\": "
```

Comment:
Would it make sense to also define a macro for the indentation level ?

Reply:
Sure, added